### DISCUSSION TITLE
No completions with some packages?

### DISCUSSION THREAD

**Discussion by Cerebus** [microsoft/pyright#4205] [Q&A] [answered]
I get completion failures with some packages (e.g., `networkx`, `scipy`, and others) but it works with other packages installed in the same venv.  I know pyright is resolving the packages correctly--I can get completions of *modules*--but completions on anything *in* the modules fails.  Further, running pyright CLI parses everything out and only complains about missing stubs.

AFAICT, all I'm getting from pyright in server mode is nil results.  Using emacs+eglot, completions at point give me:

```
[client-request] (id:383) Fri Nov 18 12:26:38 2022:
(:jsonrpc "2.0" :id 383 :method "textDocument/signatureHelp" :params
          (:textDocument
           (:uri "file:///Users/tmiller/Documents/projects/fizzy/network-resiliency/networkresiliency/graph.py")
           :position
           (:line 35 :character 2)))
[client-request] (id:384) Fri Nov 18 12:26:38 2022:
(:jsonrpc "2.0" :id 384 :method "textDocument/hover" :params
          (:textDocument
           (:uri "file:///Users/tmiller/Documents/projects/fizzy/network-resiliency/networkresiliency/graph.py")
           :position
           (:line 35 :character 2)))
[client-request] (id:385) Fri Nov 18 12:26:38 2022:
(:jsonrpc "2.0" :id 385 :method "textDocument/documentHighlight" :params
          (:textDocument
           (:uri "file:///Users/tmiller/Documents/projects/fizzy/network-resiliency/networkresiliency/graph.py")
           :position
           (:line 35 :character 2)))
[server-reply] (id:383) Fri Nov 18 12:26:38 2022:
(:jsonrpc "2.0" :id 383 :result nil)
[server-reply] (id:384) Fri Nov 18 12:26:38 2022:
(:jsonrpc "2.0" :id 384 :result nil)
[server-reply] (id:385) Fri Nov 18 12:26:38 2022:
(:jsonrpc "2.0" :id 385 :result nil)
```

However, `pylsp` running with emacs+eglot installed into the same project venv works correctly (if far slower).

**Comment by erictraut** [comment, marked as answer]
Completion suggestions are based on type information. If a library claims not to be typed and there are no stubs available, pyright by default treats all symbols exported from that library as an `Unknown` type. That means no completions will be available. If you want completions, you have a couple of options. First, you can find and install type stubs for the libraries of interest. Pylance ships with some partial stubs for popular libraries like scipy. You can find them [here](https://github.com/microsoft/python-type-stubs). In addition, you can configure pyright to attempt to infer types for library code by enabling the `useLibraryCodeForTypes` [configuration option](https://github.com/microsoft/pyright/blob/main/docs/configuration.md). Inferred return types are often incomplete and sometimes inaccurate, so they are not recommended for use with static type checking. But they're usually sufficient for completion suggestions.
